src/writer/normalized.rs

```diff
@@ -0,0 +1,549 @@
+                Feature::Rule(r, ev) => match ev {
+                    Rule::Started => self.queue.new_rule(&f, r),
+                    Rule::Scenario(s, ev) => {
+                        self.queue.insert_scenario_event(&f, Some(r), s, ev);
+                    }
+                    Rule::Finished => self.queue.rule_finished(&f, r),
+                },
+            },
+        }
+
+        while let Some(feature_to_remove) =
+            self.queue.emit((), &mut self.writer).await
+        {
+            self.queue.remove(&feature_to_remove);
+        }
+
+        if self.queue.is_finished() {
+            self.writer.handle_event(Cucumber::Finished).await;
+        }
+    }
+}
+
+#[async_trait(?Send)]
+impl<'val, W, Wr, Out> OutputtedWriter<'val, W, Out> for Normalized<W, Wr>
+where
+    Wr: OutputtedWriter<'val, W, Out>,
+    Out: 'val,
+{
+    async fn write(&mut self, val: Out)
+    where
+        'val: 'async_trait,
+    {
+        self.writer.write(val).await;
+    }
+}
+
+/// Normalization queue for incoming events.
+///
+/// We use [`LinkedHashMap`] everywhere throughout this module to ensure FIFO
+/// queue for our events. This means by calling [`next()`] we reliably get
+/// the currently outputting item's events. We're doing that until it yields
+/// event that corresponds to item being finished, after which we remove the
+/// current item, as all its events have been printed out and we should do it
+/// all over again with a [`next()`] item.
+///
+/// [`next()`]: std::iter::Iterator::next()
+#[derive(Debug)]
+struct Queue<K: Eq + Hash, V> {
+    started_emitted: bool,
+    queue: LinkedHashMap<K, V>,
+    finished: bool,
+}
+
+impl<K: Eq + Hash, V> Queue<K, V> {
+    /// Creates a new [`Queue`].
+    fn new() -> Self {
+        Self {
+            started_emitted: false,
+            queue: LinkedHashMap::new(),
+            finished: false,
+        }
+    }
+
+    /// Marks that [`Queue`]'s started event was emitted.
+    fn started_emitted(&mut self) {
+        self.started_emitted = true;
+    }
+
+    /// Checks if [`Queue`]'s started event was emitted.
+    fn is_started_emitted(&self) -> bool {
+        self.started_emitted
+    }
+
+    /// Marks this [`Queue`] as finished.
+    fn finished(&mut self) {
+        self.finished = true;
+    }
+
+    /// Checks whether [`Queue`] has been received.
+    fn is_finished(&self) -> bool {
+        self.finished
+    }
+
+    /// Removes key from the [`Queue`].
+    fn remove(&mut self, key: &K) {
+        drop(self.queue.remove(key));
+    }
+}
+
+/// [`Queue`] which can remember current item ([`Feature`], [`Rule`],
+/// [`Scenario`] or [`Step`]) and pass events connected to it to the underlying
+/// `writer`.
+///
+/// [`Feature`]: gherkin::Feature
+/// [`Rule`]: gherkin::Rule
+/// [`Scenario`]: gherkin::Scenario
+/// [`Step`]: gherkin::Step
+#[async_trait(?Send)]
+trait Emit<World> {
+    type Current;
+    type Emitted;
+    type EmittedPath;
```

Comment:
Documentation is quite lacking here, so doesn't allow to understand what is going on here.